### PR TITLE
Tree converter improving

### PR DESCRIPTION
**Pull request by elena-lyulina**
- improve implementation of tree converter
- add post-order numbering for psi trees 
- create a new architecture for Numbering
- add tests for tree conversion with post-order numbering

### CODE REVIEW

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverterTest.kt [hunk 1]
```diff
@@ -13,31 +13,35 @@ import org.jetbrains.research.ml.coding.assistant.util.PsiTestUtil.equalTreeStru
...
+    @Parameterized.Parameter(1)
+    var numbering: Numbering? = null
+
     @Test
     fun `converting PSI to GumTree tree test`() {
         val inFilePsi = myFixture.configureByFile(inFile!!.name)
         val inContext = ApplicationManager.getApplication().runReadAction<TreeContext> {
-            inFilePsi.preOrderNumbering()
-            TreeConverter.convertTree(inFilePsi)
+            PsiTreeConverter.convertTree(inFilePsi, numbering!!)
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverterTest.kt:44]: I can not comment on the 46 line, but maybe should we send the tree order to the function? Because now it compares two trees according to preOrder traversal. What do you think? Or is it not necessary? What do you think?

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/util/PsiTestUtil.kt [hunk 2]
```diff
@@ -47,7 +45,7 @@ object PsiTestUtil {
             return false
         }
         // Compare labels
-        if (psi.label != tree.label) {
+        if (psi.getLabel() != tree.label) {
```
Comment (nbirillo) [inline src/test/kotlin/org/jetbrains/research/ml/coding/assistant/util/PsiTestUtil.kt]: Why did you change the property to the function? I think the old version is better...

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverter.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            val parent = gumTreeParents.poll()
+            psiChildren.poll().forEach {
+                val tree = context.createTree(it)
+                tree.setParentAndUpdateChildren(parent)
+                gumTreeParents.add(tree)
+                psiChildren.add(it.children.toList())
+            }
+        }
+
+        numbering?.number(psiRoot, context)
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverter.kt]: Sorry, but I am not sure about this. Because we have to number the context in any cases - gumtree use ids for actions for example. Moreover, you should run `context.root.refresh()` for `TreeContext` before numbering.

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                    t.putCopyableUserData(psiId, i)
+                }
+            }
+        }
+    }
+
+    protected abstract fun PsiElement.iterable(): Iterable<PsiElement>
+    protected abstract fun ITree.iterable(): Iterable<ITree>
+
+    fun number(psiElement: PsiElement) {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt:35]: Should we make it `private`?

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        PsiTreeUtils.numbering(psiElement.iterable())
+    }
+
+    fun number(gumTreeContext: TreeContext) {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt:39]: Should we make it `private`?

**Review by nbirillo — CHANGES_REQUESTED** [review]
Almost done, thanks! But I have a few comments

File: src/test/kotlin/org/jetbrains/research/ml/coding/assistant/util/PsiTestUtil.kt (hunk 2 above)
Comment (elena-lyulina) [reply src/test/kotlin/org/jetbrains/research/ml/coding/assistant/util/PsiTestUtil.kt]: Because it was a field inside the class, and now we don't have such a class
I'm not sure we want to have a global property with a label...

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverter.kt (hunk 3 above)
Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverter.kt]: Ok, let's change it to the 'not null' type. And I do run context.root.refresh() before numbering, not quite sure why you mentioned it here

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt (hunk 4 above)
Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt:35]: I thought we want to have a possibility to number only one tree

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt (hunk 5 above)
Comment (elena-lyulina) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt:39]: I thought we want to have a possibility to number only one tree

File: src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverter.kt (hunk 3 above)
Comment (nbirillo) [reply src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/PsiTreeConverter.kt]: Oh sorry, it is my mistake

**Review by nbirillo — APPROVED** [review]
Great, thanks!
